### PR TITLE
ZeroExSwap for weth to wstEth swaps by updating BaseV2Vault code

### PR DESCRIPTION
**Pull request by dxganta**
It was not possible to use zeroExSwap by the backend to swap from weth to wstEth due to certain restrictions within the code created for security purposes. But we need zeroExSwap for weth=>wstEth  for access to better exchange rates.

- Changes to the zeroExSwap method in BaseV2Vault contract. Adding a tokenOut parameter and replacing asset with tokenOut.
- Also added a whitelist for tokenOuts during zeroExSwap for security.

### CODE REVIEW

File: src/steth/BaseV2Vault.sol [hunk 1]
```diff
@@ -133,19 +156,30 @@ abstract contract BaseV2Vault is sc4626, IFlashLoanRecipient {
...
+    function zeroExSwap(
+        ERC20 _tokenIn,
+        ERC20 _tokenOut,
+        uint256 _amount,
+        bytes calldata _swapData,
+        uint256 _assetAmountOutMin
+    ) external {
         _onlyKeeperOrFlashLoan();
 
+        if (!zeroExSwapWhitelist[_tokenOut]) revert InvalidOutToken();
```
Comment (brkomir) [inline src/steth/BaseV2Vault.sol:181]: maybe say something like "not allowed" and pass the token address?

File: src/steth/scWETHv2.sol [hunk 2]
```diff
@@ -60,7 +60,9 @@ contract scWETHv2 is BaseV2Vault {
 
     constructor(address _admin, address _keeper, WETH _weth, Swapper _swapper, PriceConverter _priceConverter)
         BaseV2Vault(_admin, _keeper, _weth, _priceConverter, _swapper, "Sandclock WETH Vault v2", "scWETHv2")
-    {}
+    {
+        zeroExSwapWhitelist[ERC20(C.WSTETH)] = true;
```
Comment (brkomir) [inline src/steth/scWETHv2.sol:64]: what about weth?

**Review by brkomir — COMMENTED** [review]
tests are lacking, and tbh i don't fully agree about adding this whitelisting functionality, I don't think it's necessary to add more functionality just before the release.

**Comment by dxganta** [comment]
> tests are lacking, and tbh i don't fully agree about adding this whitelisting functionality, I don't think it's necessary to add more functionality just before the release.

I disagree that its a functionality. Its more of a security measure.

File: src/steth/scWETHv2.sol (hunk 2 above)
Comment (dxganta) [reply src/steth/scWETHv2.sol:64]: Its added in the constructor of the BaseV2Vault contract.

File: src/steth/BaseV2Vault.sol [hunk 3]
```diff
@@ -54,6 +60,23 @@ abstract contract BaseV2Vault is sc4626, IFlashLoanRecipient {
...
+
+    /**
+     * @notice whitelist (or cancel whitelist) a token to be swapped out using zeroExSwap
+     * @param _token The token to whitelist
+     * @param _value whether to whitelist or cancel whitelist
+     */
+    function whiteListOutToken(ERC20 _token, bool _value) external {
+        _onlyAdmin();
+
+        if (address(_token) == address(0)) revert ZeroAddress();
```
Comment (brkomir) [inline src/steth/BaseV2Vault.sol:75]: add test for the "zero address" case

**Review by brkomir — COMMENTED** [review]
you have a point in the previous comment. Add the missing test and it's good to go

File: src/steth/BaseV2Vault.sol (hunk 3 above)
Comment (dxganta) [reply src/steth/BaseV2Vault.sol:75]: [Added](https://github.com/lindy-labs/sandclock-contracts/pull/120/commits/311c30f7c5dde9b770b3c00ae59c0ed55c982870)

File: src/steth/BaseV2Vault.sol (hunk 1 above)
Comment (dxganta) [reply src/steth/BaseV2Vault.sol]: [Chaged](https://github.com/lindy-labs/sandclock-contracts/pull/120/commits/311c30f7c5dde9b770b3c00ae59c0ed55c982870)

**Comment by dxganta** [comment]
@brkomir I have added a test for the zero Address case. Did you mean for any other tests that I should add here?

**Comment by brkomir** [comment]
> @brkomir I have added a test for the zero Address case. Did you mean for any other tests that I should add here?

not really, the `test_whiteListOutToken` covers adding/removing from whitelist so it's ok, although it could have been structured a little bit better, like having two separate tests. One special case I was thinking about is that we shouldn't be able to remove the base asset from the whitelist

File: src/errors/scErrors.sol [hunk 4]
```diff
@@ -21,3 +21,4 @@ error FlashLoanAmountZero();
 error ProtocolNotSupported(uint256 adapterId);
 error ProtocolInUse(uint256 adapterId);
 error FloatBalanceTooLow(uint256 actual, uint256 required);
+error TokenNotAllowed(address token);
```
Comment (brkomir) [inline src/errors/scErrors.sol:24]: nitpicking but would be more informative to name it `TokenOutNotAllowed`

Comment (dxganta) [reply src/errors/scErrors.sol]: Agree. [Changed](https://github.com/lindy-labs/sandclock-contracts/pull/120/commits/c62ab80b2a07e7215129dd106bafac6d891c31e9).

**Comment by dxganta** [comment]
> > @brkomir I have added a test for the zero Address case. Did you mean for any other tests that I should add here?
> 
> not really, the `test_whiteListOutToken` covers adding/removing from whitelist so it's ok, although it could have been structured a little bit better, like having two separate tests. One special case I was thinking about is that we shouldn't be able to remove the base asset from the whitelist

I can't think of a scenario when the Admin address would do that other than if the Admin address gets compromised in which case it can do al lot other things to compromise the vault.

**Review by brkomir — APPROVED** [review]
